STACpopulator/extensions/base.py

```diff
@@ -0,0 +1,360 @@
+    ServiceType,
+    ncattrs_to_bbox,
+    ncattrs_to_geometry,
+)
+import types
+from STACpopulator.extensions.datacube import DataCubeHelper
+from STACpopulator.extensions.thredds import THREDDSHelper
+
+T = TypeVar("T", pystac.Collection, pystac.Item, pystac.Asset, item_assets.AssetDefinition)
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ExtensionHelper(BaseModel):
+    """Base class for dataset properties going into the catalog.
+
+    Subclass this with attributes.
+
+    Attributes
+    ----------
+    _prefix : str
+        If not None, this prefix is added to ingested data before the jsonschema validation.
+    _schema_uri : str
+        URI of the json schema to validate against.
+    _schema_exclude : list[str]
+        Properties not meant to be validated by json schema, but still included in the data model.
+    """
+    _prefix: str = PrivateAttr()
+    _schema_uri: FilePath = PrivateAttr(None)
+    _schema_exclude: list[str] = PrivateAttr([])
+
+    model_config = ConfigDict(populate_by_name=True, extra="ignore")
+
+    @model_validator(mode="before")
+    @classmethod
+    def validate_jsonschema(cls, data):
+        """Validate the data model against the json schema, if given."""
+        # Load schema
+        uri =  cls._schema_uri.default
+        if uri is not None:
+            schema = json.load(open(uri))
+            validator_cls = jsonschema.validators.validator_for(schema)
+            validator_cls.check_schema(schema)
+            validator = validator_cls(schema)
+
+            attrs = {f"{cls._prefix.default}:{k}": v for (k,v) in data.items() if k not in cls._schema_exclude.default}
+            errors = list(validator.iter_errors(attrs))
+            if errors:
+                raise ValueError(errors)
+
+        return data
+
+    def apply(self, item, add_if_missing=False):
+        """Add extension for the properties of the dataset to the STAC item.
+        The extension class is created dynamically from the properties.
+        """
+        ExtSubCls = metacls_extension(self._prefix, schema_uri=str(self._schema_uri))
+        item_ext = ExtSubCls.ext(item, add_if_missing=add_if_missing)
+        item_ext.apply(self.model_dump(mode="json", by_alias=True))
+        return item
+
+
+class BaseSTAC(BaseModel):
+    """Base class for STAC item data models.
+
+    Attributes
+    ----------
+    geometry : AnyGeometry
+        The geometry of the dataset.
+    bbox : list[float]
+        The bounding box of the dataset.
+    start_datetime : datetime
+        The start datetime of the dataset.
+    end_datetime : datetime
+        The end datetime of the dataset.
+    extensions : list[str]
+        Name of the class attributes that point to STAC extension helper classes. Those extension classes should have an `apply` method.
+    """
+    # STAC item properties
+    geometry: AnyGeometry | None
+    bbox: list[float]
+    start_datetime: datetime
+    end_datetime: datetime
+
+    extensions: list = []
```

Comment:
Is there a reason these can't be a list of the extensions themselves? Why have this as a list of strings referring to class methods?

Reply:
I think it is better to let the helpers add them automatically, as they might change over time (e.g.: new extension version with modified attributes).

Reply:
Sorry, I'll try to explain a bit better (or maybe I'm not understanding the issue). 

Right now we do something like this:

```python
class BaseSTAC:
    ...
    def stac_item(self) -> "pystac.Item":
        ...
        for ext in self.extensions:
            getattr(self, ext).apply(item)

class THREDDSCatalogDataModel(BaseSTAC):
    ...
    properties: ExtensionHelper
    datacube: DataCubeHelper
    thredds: THREDDSHelper
    ...
    extensions: list = ["properties", "datacube", "thredds"]
```

Why can't we do this?

```python
class BaseSTAC:
    ...
    def stac_item(self) -> "pystac.Item":
        ...
        for ext in self.extensions:
            ext.apply(item)

class THREDDSCatalogDataModel(BaseSTAC):
    ...
    extensions: list = [ExtensionHelper, DataCubeHelper, THREDDSHelper]
```

Reply:
Because of data ingestion. 

For example, the THREDDS extension needs to be passed `THREDDSHelper(data['data']["access_urls"])`

I currently manage this with `model_validator`: 
```
    @model_validator(mode="before")
    @classmethod
    def thredds_helper(cls, data):
        """Instantiate the THREDDSHelper."""
        data["thredds"] = THREDDSHelper(data['data']["access_urls"])
        return data
```
I'm not sure how we'd do that with your proposal without adding some obscure magic. 
What I've now done is automatically detect extensions from the annotation (if it's a Helper subclass). Hope that's ok for now.

Reply:
I was actually misinterpreting the question thinking of `extensions` as the `STAC.Item.extensions` (ie: the URI to the applied extensions). I think we should rename the attribute, because that is very confusing. It should be `helpers` to highlight the use of the helpers that have extended capabilities for applying the STAC extensions (and sometimes non-extension attributes). 

Ideally, we should have something like: 
```python
helpers: list[Type[Helper]] = [ExtensionHelper, DataCubeHelper, THREDDSHelper]
```

Because only classes of the helpers are used (not instances), they should be able to receive the item data dynamically for the `apply()` call.

Reply:
Renamed `_extensions` to `_helpers`. 

I don't see how I could implement your proposal, without hard-coding the `data` ingestion logic into the helpers themselves, which would couple them tightly with the `Loader`, which I thought we should avoid.

Reply:
Minor nitpick. 

I think the `BaseSTAC` class should define something like: 

```python
class BaseSTAC(abc.ABC):
   @classmethod
   @abc.abstractmethod
   def helpers(cls) -> list[Type[Helper]]:
       raise NotImplementedError
```

This way, any derived class and smart IDE flags right away that `helpers` must be overridden.

I think `helpers` should be used instead of `_helpers` because it is part of the "public" interface of that class, for anyone that derives a new implementation from it.